### DISCUSSION TITLE
Add a time expansion flag to bat_ident.py

### DISCUSSION THREAD

**Discussion by Jeff37** [rdz-oss/BattyBirdNET-Analyzer#3] [Ideas]
Hi, it would be great to have the possibility to process high-frequency sound recordings that are 10-timed expanded. Often, wav files are modified for example from 144kHz to 14.4kHz for the sake of the analysis of echolocation signals.
To his API batdetect2, [@macaodha](https://github.com/macaodha/batdetect2) has added a flag "--time_expansion_factor" that is taking the value 1 or 10.
Is it possible to add such a flag in bat_indent.py ?

Thanks for your work!
JF Godeau

**Comment by rdz-oss** [comment]
Hi! Thank you for your interest and suggestion. I think that would be a worthwhile feature. 
I will put it on the feature list. To do it, there would be several options. Extend the audio.py with a function or use ffmpeg or sox. Sox is already used to plot spectrograms (python3 bat_ident.py --spectrum on) using https://github.com/rdz-oss/BattyBirdNET-Analyzer/blob/main/batchspec.py . That does require the user to install sox, however.

Until then, if you or anyone reading this have sox or sox.exe installed, you can quickly transform the file, e.g.
'sox imput.wav output.wav speed 10'
should reduce the expansion by factor 10. Same can be done with the UI of audacity, 'Effects - Pitch and tempo - Adjust speed'.


BR
Richard

**Reply by Jeff37** [reply]
Dear Richard, happy to read that the suggestion makes sense. I didn't read thoroughly your code yet, and I should admit that I'm more in R than in python, therefore I'm a bit more clumsy with understanding the codes and the functions (well, it just takes time!). I'll check sox, thanks for the tip.
So far I use this chunk of R code to transform the time expansion factor, thanks to the package 'seewave' it's simple and quite fast.

`library(seewave)`
`path4trans <- "/path/to/your/Folder"`
`LF <- list.files(path4trans, pattern = ".wav$", full.names = TRUE)`
`for(i in LF){`
`  Wv <- tuneR::readWave(i)`
`  cat(paste0(Wv@samp.rate, "\t"))`
`  savewav(Wv, f= Wv@samp.rate * 10, filename = paste(path4trans, paste0("", basename(i)), sep = "/"))`
`}`
Keeping the last paste0(...) as it is will replace the original files. Otherwise, one can simply write the transformed files in another subfolder named in the `paste0("Subfolder", basename(i), sep = "/")`
NB. this code works under Linux, for Windows users, consider changing the '/' appropriately ! 

This way, I could batch the transformation of all my wav files but since I've hundreds of Go I'm working folders per folders so far.
I 'll be patient enough to see the new feature appearing on GitHub. BTW, I saw that you updated the repo and that you added your training data set. That's awesome!! Thanks for that! Just in case, if I could contribute to add some recordings of species that are underrepresented, I'd be more than happy.
I'm mainly studying bats in Belgium and Poland but I also do have many data from Lithuania and Latvia, and some from Slovakia. For instance, I can provide you with samples of _E. nilssonii, P. nathusii, B. barbastella_, ... let me know.

I've been trying bat_ident.py on few data sets of mine and I check the results in comparison to batDetect2 and/or my own hand validation. I played around a bit with the parameters (eg. overlap, min_conf, building my own species list) and the results can be really good! In some cases (mainly when the quality of the recordings is poor) there are many false positive (wrong ID with high confidence level) then it's needed to check the sounds. Altogether, it's a really good (and fast!!) tool that you made. Thanks again for that and for sharing.

I'll send along a short report on my classification results when I'll find a bit of time to do it.

Keep in touch.
JF Godeau

**Comment by rdz-oss** [comment]
Hi JF,
thank you for the feedback. I will write you an email with some more info later tonight.
When I transform the data for the training, I use batch and python scripts to iterate through folder hierarchies. Perhaps something like that works for your batch transformations, too. It goes through all subdirectories (1 level deep) and transforms the files to 10 times speed to new folders with the same name. You can save it as e.g match_speed.sh , adapt to your needs. Before execution  ./match_speed.sh , you will need to run chmod +x match_speed.sh   
You can (commented out) also resample to another rate and cut off all sounds lower or higher than a given frequency.

BR
Richard
``` sh
 #! /bin/bash

train_dir=training_set
rm -rf "$train_dir"
mkdir "$train_dir"


FILES=*
for f in $FILES
do
  cd "$f" 
  mkdir ../"$train_dir"/"$f"
  
  for call in *.wav
  do
    sox "$call" ../"$train_dir"/"$f"/"$call" speed 10
    #  sox "$call" ../"$train_dir"/"$f"/"$call" speed 10 rate 256000 sinc -n 32767 12000-65000
  done
  cd ..
done
```
